Remove SnapshotPageProxy hard coded dependancy

FabienD opened the pull request:
According what I've explained in the issue #551, I've built this PR in order to make SnapshotPageProxy customisable in an easy way (I think).

core23 commented on Entity/Transformer.php:
```diff
@@ -33,6 +32,8 @@ class Transformer implements TransformerInterface
 
     protected $blockManager;
 
+    protected $snapshotPageProxyClass;
```
phpdoc missing

core23 commented on Entity/SnapshotManager.php:
```diff
@@ -36,6 +35,8 @@ class SnapshotManager extends BaseEntityManager implements SnapshotManagerInterf
      */
     protected $templates = array();
 
+    protected $snapshotPageProxyClass;
```
phpdoc missing

OskarStark commented on Entity/SnapshotManager.php:
```diff
@@ -43,11 +44,12 @@ class SnapshotManager extends BaseEntityManager implements SnapshotManagerInterf
      * @param array           $templates An array of templates
      */
-    public function __construct($class, ManagerRegistry $registry, $templates = array())
+    public function __construct($class, $snapshotPageProxyClass, ManagerRegistry $registry, $templates = array())
```
This is BC

Soullivaneuh commented on CmsManager/CmsSnapshotManager.php:
```diff
@@ -36,13 +35,20 @@ class CmsSnapshotManager extends BaseCmsPageManager
     {
         $this->snapshotManager = $snapshotManager;
         $this->transformer = $transformer;
+        $this->snapshotPageProxyClass = $snapshotPageProxyClass;
```
Is this should be mandatory?

In this case, you should throw a deprecated error.

FabienD replied:
Yes, it is. A default value is provided in the constructor.

Soullivaneuh commented on Model/SnapshotPageProxyInterface.php:
```diff
@@ -0,0 +1,21 @@
+ */
+interface SnapshotPageProxyInterface extends PageInterface, \Serializable
+{
+}
```
Maybe some methods could be added here?

FabienD replied:
I've done that in order to test the type of the object in "Twig/Extension/PageExtension.php"

Soullivaneuh commented:
> I've built this PR in order to make SnapshotPageProxy customisable in an easy way (I think).

Could you please elaborate? Do you have some concrete cases where this class should be extended / overridden?

FabienD commented:
Hi,

I've made this PR, just because, it's the only way I found to add a new property on the Page entity (cf. #551). I want to have the possibility to set a specific css class by page.

Thx.

Soullivaneuh commented:
@rande what do you think?

FabienD commented:
No feedback ?
Thx.

rande commented:
I will prefer a factory to avoid injecting class name.

FabienD commented:
Ok, I'll try to do that. Thank you.

FabienD commented:
Hi,

I've refactored the PR in order to use a factory to instantiate the **SnapshotProxy** class. Now, the **SnapshotManager** takes a 4th argument, the **SnapshotProxyFactory**. I'm not confortable with "tests", I've updated the **SnapshotManagerTest** but I don't know how to fix the failed test in **CmsSnapshotManagerTest**. I need help to fix the failed tests.

In addition, I've removed the **getPageByName** method from the **SnapshotManager**, this method seems not to be used, the **snapshotProxy** is instantiated with only 2 args, the class needs 3.

What do you think about this refactoring ?

Thx.

core23 commented on Entity/SnapshotManager.php:
```diff
@@ -37,17 +39,24 @@ class SnapshotManager extends BaseEntityManager implements SnapshotManagerInterf
+     * @param SnapshotPageProxyFactory $snapshotPageProxyFactory SnapshotPageProxy factory
      */
-    public function __construct($class, ManagerRegistry $registry, $templates = array())
+    public function __construct($class, ManagerRegistry $registry, $templates = array(), SnapshotPageProxyFactory $snapshotPageProxyFactory)
```
This is a BC break

core23 commented on Entity/SnapshotManager.php:
```diff
@@ -140,34 +149,6 @@ public function findEnableSnapshot(array $criteria)
-     *
-     * @return PageInterface|false
-     */
-    public function getPageByName($routeName)
```
BC break

core23 commented on Entity/Transformer.php:
```diff
@@ -52,13 +51,14 @@ class Transformer implements TransformerInterface
      * @param BlockManagerInterface    $blockManager
      * @param RegistryInterface        $registry
+     * @param string                   $snapshotPageProxyClass Namespace of SnapshotPageProxy class
      */
```
Parameter is missing in the method header ;)

Soullivaneuh commented:
According to the new [Sonata version management](https://github.com/sonata-project/SonataAdminBundle/issues/3053) and next major [release plan](https://github.com/sonata-project/SonataAdminBundle/issues/3731), this project has been refactored regarding branching and versioning.

If you see this message, your PR concerns a patch or a minor release and is not targeting the right branch.

So I'm closing this one, but don't see it as a refusal. If you think your work is still relevant and want to continue, feel free to reopen it on the right branch (e.g. the default one).

Regards.
